### ISSUE TITLE
Circular dependency resulting in infinite loop

### ISSUE THREAD

**Issue by wjohnsto** [pahen/madge#159]
I am working on a server with roughly 400 files, using GraphQL which lends itself to circular dependencies. I'm trying to use madge to prevent this as best I can, but I'm running into an issue where it appears madge ends up in an infinite loop. I've traced the issue, and fixed it for my purposes so I wanted to bring it to light here and see if it's a bug or if I'm doing something wrong.

My project is structured so that I have a bunch of `component` directories, each with an `index.js` that exports all the modules in the component. So something like this:

```
- component
  - controller.js
  - model.js
  - index.js
  - graphql
    - type.js
    - mutation.js
    - query.js
    - index.js
```

I've traced the issue down to [tree.js#L169](https://github.com/pahen/madge/blob/master/lib/tree.js#L169). It appears in the case of my project `depTree` (which is the result of the call to `dependency-tree` has in itself circular references. I'm not sure if this is intended by `dependency-tree`, but I'm assuming it is. This would be fine if the `convertTree` function had a short-circuit in the event that there was a circular reference, but I can't find where it does that. I'm assuming that it is intended for `tree` to be that object. When I change the function to be the following it fixes the issues:

```
convertTree(depTree, tree, pathCache) {
    for (const key in depTree) {
        const id = this.processPath(key, pathCache);

        if (!tree[id]) {
            tree[id] = [];

            for (const dep in depTree[key]) {
                tree[id].push(this.processPath(dep, pathCache));
            }

            this.convertTree(depTree[key], tree, pathCache);
        }
    }

    return tree;
}
```

Basically, I only call to `convertTree` in the event that `id` isn't found in the flat `tree`. This seems like it might be the intended functionality, but I'm not sure exactly so I thought I would post here rather than make a PR.

Has anyone else run across this same issue?

**Comment by rokoroku** [comment]
It solves me too! Thank you for sharing your solution. 
(I use exactly same structure like @wjohnsto , export every module from a folder via `index.js`)

**Comment by sandkampderek** [comment]
I have this problem as well. It currently takes madge 12.5 minutes to check our codebase for circular dependencies. After making the change @wjohnsto suggested, it takes only 22 seconds. For what it's worth, we also have quite a few `index.js` files that re-export things within their folder, though I didn't verify that that was the cause of the problem.

**Comment by wjohnsto** [comment]
I do think this is the intended functionality. I've been trying to think of a case where you would want it the other way, but I'm unable to find one. Maybe @pahen understands the logic a little more, it seems like he has done the bulk of the work in tree.js.

**Comment by pahen** [comment]
Where you able to verify that the tree from `dependency-tree` had a circular reference in your project? @wjohnsto I'm curious how it looked. I'm looking at writing a test for the fix you provided ..

**Comment by wjohnsto** [comment]
I was able to verify that there were multiple circular dependencies. I can try to work up the simplest example I can, as what I have currently is a little bit complex for a gist.

**Comment by pahen** [comment]
Thanks for all your effort @wjohnsto :) I was able to verify the problem and your fix now. Good job! I've added your fix and released it in Madge v3.1.0 now.

**Comment by wjohnsto** [comment]
Awesome! 👍
